proto/.gitignore

```diff
@@ -0,0 +1,4 @@
+out/
```

Comment:
I don't think it makes sense to place that there. This is just for protobufs, it is not a node package. We are currently only generating and publishing but we will need to support other languages soon as well.

Reply:
It is totally a node package, if it wouldn't, we wouldn't have `package.json` and we wouldn't publish to npm.

`node/packages` was configured specifically for that, so e.g. Python work can live in `python` and go in `go` etc.

Reply:
This repository is organized in a way that all npm packages that are maintained in its scope are located at `node/packages` folder, we should not break from that

Reply:
I mean to build and publish python we will have a `setup.cfg` file in there as well. Go will have a `go.mod`. I understand the organization goals, but this is not strictly a node project.

We can create a folder just for the protobufs and I believe can just reference them from a build script for node under `node/packages` if you feel strongly about that organization.

Reply:
> We can create a folder just for the protobufs and I believe can just reference them from a build script for node under node/packages if you feel strongly about that organization.

Yes, it should be done that way

Reply:
Go ahead and take a look at the latest commit @medikoo , should be what you expect now.

Reply:
Thanks it looks way better. What's left is that top-level `package.json` should be removed,  and all dev dependencies as listed in `node/packages/sdk-schema` should be moved to `node/package.json`

Addtionally can you setup bundling with esbuild instead of rollup? We already use esbuild in context of this repository, and for maintenance clarity it'll be great to stick to one bundler (unless you're doing with rollup things that cannot be done with esbuild)

Reply:
> What's left is that top-level package.json should be removed,
My bad, I was looking at the `node/package.json` but running commands in the root I guess 🤦 


I'll need to look into this a bit. I'm not sure the esbuild supports what we need for this bundling. I know esbuild use to require a single entry point, but we don't actually have a single entrypoint from the generated ts files.

Reply:
I see, so it's only about transpiling all `.ts` into `.js` (?) In such case `typescript` dependency cannot do that?

Reply:
Yeah so looking at it, I think `rollup` is a lot cleaner here since it,

1. Handles the multiple entry points
2. Creates a bundled `.d.ts` file that will work with our bundled `cjs` file

If we used `typescript` directly it would create the same problem as `esbuild`,

For example after we run `buf generate` I'll have a ts file for each proto file,
1. `out/serverless/proto/instrumentation/v1/trace.pb.ts`
2. `out/serverless/proto/instrumentation/v1/metric.pb.ts`

And so on. If I use `esbuild` or `typescript`, we will get output then of,

1. `dist/serverless/proto/instrumentation/v1/trace.pb.js`
2. `dist/dts/serverless/proto/instrumentation/v1/trace.pb.d.ts`
3.  `dist/serverless/proto/instrumentation/v1/metric.pb.js`
4.  `dist/dts/serverless/proto/instrumentation/v1/metric.pb.d.ts`

Compared to `rollup` which will let us output,

1. `dist/index.cjs.js`
2. `dist/index.d.ts`

Not to mention the `esbuild`/`typescript` approach would result in imports like,
`import { TracePayload } from '@serverless/sdk-schema/serverless/instrumentation/v1/trace`
vs. from rollup we get,
`import { TracePayload} from '@serverless/sdk-schema`


So I think it is worth the it to use `rollup` in this scenario to make the best Developer friendly NPM package.

Reply:
Also just in general we should never use `typescript` directly for transpiling, it is horrendously slow. That's why we tend to introduce `esbuild` and `rollup` most places.

Reply:
Thanks for looking into the details. 

From an extension (and I think maintenance as well) perspective, it'll be nicer to have all those schemas built into individual files, assuming they internally solve their dependencies so e.g. `trace.pb.js` requires internally `tags.pb.js`, which is also required by `metrics.pb.js` - Still if I remember correctly it's not possible with `protoc` (?)

Having that in extension I can just require `trace.pb.js` (as I don't need metrics there)

Reply:
If we go that route, what would end up happening is every reference will get bundled into each one. So we will get a `common.pb.js` and `tags.pb.js` that will contain the common messages and the tag messages. But `trace.pb.js` will contain all the the `tags.pb.js` included in its own file. That also means the single external dependency will get bundled into each file instead of just once.

I'd prefer to go with a single bundle. IMO is the much cleaner approach vs having large individual files to import.

Reply:
Also this is a package that needs to be used internally for our typescript services and having the single bundle and single type declaration file makes us not have to jump through hoops to use it. From a performance perspective and best practices perspective, using `rollup` here is the right call.

Reply:
> That's not the route I had in mind. I meant scenario where we create pb.js file for each .proto file, but not as standalone bundle,

I'm sure we could get it to work this way, but I don't think that is the best route. It is optimizing for only one of the uses of this package. 

I think if you want that functionality for the extension then we should just transpile the protobuf's you need into the extension package vs. hampering all other usecases.

Reply:
Also the more I look  at `esbuild`, I don't think we can actually do what you are suggesting, it will always try to bundle the inter module references into single files. So we can mark the `protoscript` dependency as external but still will have tags and common messages added in all the other files.

Reply:
> needs to be used internally for our typescript services 

What are "typescript services" ?

> From a performance perspective and best practices perspective, using rollup here is the right call.

In my understanding, we're not creating here browser ready app, or final extension layer. We're just creating JS/TS interfaces for Proto files to be used on backed and in the extension, that's all. 

So ideally if we end here with clear modules-set for use on our ends.

Reply:
> Also the more I look at esbuild, I don't think we can actually do what you are suggesting, it will always try to bundle the inter module references into single files.

I think it's not question for `esbuild` but for tool, which we use to transpile `.proto` into `.ts`. 

Is it possible to have modules tree created with it? Where e.g. `trace.pb.ts` imports `tags.pb.ts` instead of bundling everything into single `index.ts` ?

Reply:
We use `typescript` across the entire internal platform (we should be using typescript everywhere, but that is fight for another day 😆 ). The solution you are suggesting will create complexity for us to include types as we will not have a single entry point for the library. 

We are publishing a package that will,
1. Provide interfaces for JS/TS
2. For typescript services it will enforce strong typing of our payloads
3. Provide built in functions on the interface prototypes for encoding/decoding


I wrote a solution here that will meet both usecases, I'm not seeing the point of complicating this. If we go your route we will have to create an internal package version of this just to meet our needs. If you can give me a legit reason that causes us to need the approach you are suggesting on the extension then we can deal with two split packages, but as it stands I'm not seeing why we would build this package differently from every modern JS/TS package on NPM.

Reply:
> Is it possible to have modules tree created with it? Where e.g. trace.pb.ts imports tags.pb.ts instead of bundling everything into single index.ts ?

We'd have to use `tsc` directly and then everything would then need a `peerDependency` on `protoscript` to work

Reply:
> The solution you are suggesting will create complexity for us to include types as we will not have a single entry point for the library.

Hmm.. I don't think I meant that.  We can have `traces.ts`, `metrics.ts` and if there's a need also `index.ts` that serves as an single entry point (which imports and exports both _traces_ and _metrics_) 

>  I'm not seeing why we would build this package differently from every modern JS/TS package on NPM.

I don't see where I suggest that. I'm just calling for better modularity, that's it. Again, our goal here should be to create clean JS/TS interfaces for proto files, and not bundles

Reply:
> We'd have to use tsc directly 

Why? I didn't say we should not create typings and JS files

> everything would then need a peerDependency on protoscript to work

It definitely shouldn't be the case. `protoscript` should be a dependency of this package, and should be naturally required/imported in generated TS/JS modules (proto counterparts)

Question is whether we can transpile`.proto` to `.ts` so it's setup that way (?)

Reply:
I guess, I'm misunderstanding what you expect then, but we don't need modularity if you are taking advantage of the type definitions and everything. Everything is clean if we use the tooling available to us.

We can setup a separate folder in the package that contains what you want, It would look like this,

```
npm notice === Tarball Contents ===
npm notice 380B    README.md
npm notice 259.8kB dist/index.cjs.js
npm notice 347.0kB dist/index.cjs.js.map
npm notice 43.0kB  dist/index.d.ts
npm notice 259.2kB dist/index.esm.js
npm notice 347.0kB dist/index.esm.js.map
npm notice 143.9kB dist/serverless/proto/common/v1/common.pb.js
npm notice 245.5kB dist/serverless/proto/common/v1/common.pb.js.map
npm notice 9.0kB   dist/serverless/proto/common/v1/common/v1/common.pb.d.ts
npm notice 31.4kB  dist/serverless/proto/common/v1/instrumentation/tagset/v1/tagset.pb.d.ts
npm notice 4.4kB   dist/serverless/proto/common/v1/instrumentation/v1/log.pb.d.ts
npm notice 6.5kB   dist/serverless/proto/common/v1/instrumentation/v1/metric.pb.d.ts
npm notice 4.2kB   dist/serverless/proto/common/v1/instrumentation/v1/trace.pb.d.ts
npm notice 9.0kB   dist/serverless/proto/instrumentation/v1/common/v1/common.pb.d.ts
npm notice 31.4kB  dist/serverless/proto/instrumentation/v1/instrumentation/tagset/v1/tagset.pb.d.ts
npm notice 4.4kB   dist/serverless/proto/instrumentation/v1/instrumentation/v1/log.pb.d.ts
npm notice 6.5kB   dist/serverless/proto/instrumentation/v1/instrumentation/v1/metric.pb.d.ts
npm notice 4.2kB   dist/serverless/proto/instrumentation/v1/instrumentation/v1/trace.pb.d.ts
npm notice 136.9kB dist/serverless/proto/instrumentation/v1/log.pb.js
npm notice 239.9kB dist/serverless/proto/instrumentation/v1/log.pb.js.map
npm notice 154.5kB dist/serverless/proto/instrumentation/v1/metric.pb.js
npm notice 255.5kB dist/serverless/proto/instrumentation/v1/metric.pb.js.map
npm notice 219.7kB dist/serverless/proto/instrumentation/v1/trace.pb.js
npm notice 312.8kB dist/serverless/proto/instrumentation/v1/trace.pb.js.map
```

And then your imports would be `import {} from '@serverless/sdk-schema/dist/serverless/proto/instrumentation/v1/trace.pb.js'`

I'm okay with doing this, but I still am of the opinion that the bundle makes way more sense overall.

Reply:
> but we don't need modularity if you are taking advantage of the type definitions and everything.

I'm not sure if it's like that, as if what's created here are bundles, when I require it in the extension, I'd end with _metrics_ and other schemas, which extension doesn't use, in its final bundle (not sure how tree-shaking will work on those result files).

Another argument is that it's cleaner for code inspection to have all those modules organized as multiple files and not a big single index.

Nonetheless, it's not that big deal, I don't think we should focus on that extensively now.

If it's easy for you to set this up, so it's modular, I think it'll be great to have it, otherwise, please skip it. If I'll find later that this still bothers me, I can jump and cover it in a way so your work on the backend is not affected, and the extension is happy.

Reply:
I'm fine to include the second folder that contains just the straight one to one if that is what makes the most sense. I just built a version that contains the bundle and the one-to-one version and it doesn't seem to interfere. We can go with that for now and always go a different direction if we need.

Reply:
@medikoo Here is an example usage using the pattern you suggested,

If you are okay with this I'll add it to the published package,

```javascript
const { randomUUID } = require('crypto')
// const { SpanJSON } = ServerlessSdkSchema;
const { Span } = require('@serverless/sdk-schema/dist/serverless/proto/instrumentation/v1/trace.pb')
const run = () => {
  const spanId = randomUUID();
  const traceId = randomUUID();

  const now = Date.now();

  const nanoNow = BigInt(now) * BigInt(1000000);

  const span = Span.encode({
    id: Buffer.from(spanId, 'utf-8'),
    traceId: Buffer.from(traceId, 'utf-8'),
    name: 'test',
    startTimeUnixNano: nanoNow,
    endTimeUnixNano: nanoNow,
    tags: [{
      key: 'test',
      value: 'abc123'
    }]
  })

  return span;
}
```

Reply:
In general it looks good.

I have just concern about noisy require path (ideally it should be just `@serverless/sdk-schema/trace`) - still it relates to my other comment of how we organize schema files - so we don't have to dive into it here.

Also tags structure I believe, will be different (at least in your example it doesn't seem to conform to the schema).

Reply:
> I have just concern about noisy require path (ideally it should be just @serverless/sdk-schema/trace) - still it relates to my other comment of how we organize schema files - so we don't have to dive into it here.


Yeah I haven't addressed the paths, but I can do what you mentioned I believe, let me check the config.

> Also tags structure I believe, will be different (at least in your example it doesn't seem to conform to the schema).

Yeah, I just copied an old example for this, didn't actually run it. It would fail to run in this example cause of that tags version.

Reply:
@medikoo , Here is what I think is the finalized folder structure (minus moving `tagsets` to just `tags`),

```
npm notice 📦  @serverless/sdk-schema@0.0.0
npm notice === Tarball Contents ===
npm notice 379B    README.md
npm notice 16.9kB  dist/common/v1/common.pb.js
npm notice 16.2kB  dist/common/v1/common.pb.js.map
npm notice 123.5kB dist/index.cjs.js
npm notice 123.3kB dist/index.cjs.js.map
npm notice 55.1kB  dist/index.d.ts
npm notice 121.1kB dist/index.esm.js
npm notice 123.1kB dist/index.esm.js.map
npm notice 78.6kB  dist/instrumentation/tagset/v1/tagset.pb.js
npm notice 78.5kB  dist/instrumentation/tagset/v1/tagset.pb.js.map
npm notice 10.7kB  dist/instrumentation/v1/log.pb.js
npm notice 10.6kB  dist/instrumentation/v1/log.pb.js.map
npm notice 25.7kB  dist/instrumentation/v1/metric.pb.js
npm notice 25.7kB  dist/instrumentation/v1/metric.pb.js.map
npm notice 81.8kB  dist/instrumentation/v1/trace.pb.js
npm notice 83.0kB  dist/instrumentation/v1/trace.pb.js.map
```

You good with this?

Reply:
I would remove also `v1`, as I don't think we will ever host two different versions of schemas (not mentioning definining v2)

Also why `instrumentation` ?

Can't it be that `traces`, `metrics` etc. are maintained  top level ?

Reply:
Alright, we can remove the `v1` and the `instrumentation`, 

```
npm notice 📦  @serverless/sdk-schema@0.0.0
npm notice === Tarball Contents ===
npm notice 379B    README.md
npm notice 19.5kB  dist/common/common.pb.js
npm notice 16.2kB  dist/common/common.pb.js.map
npm notice 144.2kB dist/index.cjs.js
npm notice 123.3kB dist/index.cjs.js.map
npm notice 55.1kB  dist/index.d.ts
npm notice 142.0kB dist/index.esm.js
npm notice 123.1kB dist/index.esm.js.map
npm notice 12.5kB  dist/log.pb.js
npm notice 10.5kB  dist/log.pb.js.map
npm notice 30.1kB  dist/metric.pb.js
npm notice 25.7kB  dist/metric.pb.js.map
npm notice 91.8kB  dist/tagset/tagset.pb.js
npm notice 78.5kB  dist/tagset/tagset.pb.js.map
npm notice 95.6kB  dist/trace.pb.js
npm notice 83.0kB  dist/trace.pb.js.map
```

Reply:
Looks great, (just `dist/tagset/tagset.pb.js` -> `dist/tags.pb.js`) 👍

Reply:
@medikoo yup, working on the tag changes next